text/0061-minimal-streams.md

```diff
@@ -0,0 +1,275 @@
+
+Amaranth supports specifying the structure of such interfaces with the very general `amaranth.lib.wiring` module, but the structure is only half of the story, as even the simplest flow control schemes include complex requirements, restricting not only the permissible sequences of states of flow control signals, but often also the domain from which the signals may be driven, and which signals may combinatorially depend on which other ones. A mistake in flow control circuitry is often difficult to discover (it may only manifest when used with a subset of compatible blocks and not surface in tests) and debug (waveform viewers generally do not show data dependencies so it may not be clear why a certain signal is asserted or deasserted).
+
+This complexity and challenges makes it useful to standardize on specific data exchange interfaces with a simple, robust, flexible, and well-defined flow control scheme. This proposal offers one such scheme suited for unidirectional communication only. In turn, it can be used to build bidirectional communication protocols.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `amaranth.lib.stream` module defines the concept of a *stream*. A stream is an `amaranth.lib.wiring` interface consisting of three port members:
+- `payload: Out(payload_shape)`
+- `valid: Out(1)`
+- `ready: In(1)`
+
+The purpose of a stream is to mediate unidirectional data transfer between two components, called a *transmitter component* (which drives `payload` and `valid` members in its *output stream*, and samples `ready`) and a *receiver component* (which drives `ready` member in its *input stream*, and samples `payload` and `valid`). The directions of port members above are from the perspective of the transmitter component, as required by `amaranth.lib.wiring`. The values of every member must be fixed or change according to the same clock domain.
+
+Data is transferred via the stream according to the following rules:
+1. On each cycle where both `valid` and `ready` are asserted, a transfer is performed: the contents of `payload` is conveyed from the transmitter to the receiver.
+2. Once the transmitter asserts `valid`, it must not deassert it until a transfer is performed.
+3. As an exception to rule (2), when the transmitter is reset it must deassert `valid`, unless it ties `valid` to `Const(1)` in the interface object.
+4. Once the transmitter asserts `valid`, it must not change the contents of `payload` until a transfer is performed.
+5. The transmitter must not assert `valid` in response to the receiver asserting `ready` (combinatorially or not).
+6. Once the receiver asserts `ready` it may deassert it at any time.
+7. The receiver may assert `ready` in response to the transmitter asserting `valid` (including combinatorially).
```

Comment:
I found rules 5 and 7 a little difficult to parse, although I believe I understand the intent. Some possible alternate formulations:
5. The transmitter must not condition the assertion of `valid` on the state of `ready` from the receiver
7. The receiver may condition the state of `ready` on the state of `valid` (or the contents of `payload`)(including combinatorially) 
(In addition to the previously discussed case of buffer allocation, another case where  ready may depend on both valid and data would be a block implementing routing or demultiplexing based on a `tdest` type signal in the payload)

Reply:
"Condition the assertion" and "condition the state" are more confusing to me (introduces a new term used nowhere else in the spec; also I don't generally recall reading a lot of text using this terminology).

Reply:
AXI-Stream uses this wording:

> A Transmitter is not permitted to wait until TREADY is asserted before asserting TVALID. Once TVALID is
asserted, it must remain asserted until the handshake occurs.
>
> A Receiver is permitted to wait for TVALID to be asserted before asserting TREADY. It is permitted that a
Receiver asserts and deasserts TREADY without TVALID being asserted.

Maybe "wait for" is clearer than "in response to" or "condition". The main reason why these conditions are in place is to avoid a deadlock if the producer waits for `ready` and the consumer waits for `valid`, so one of them is not supposed to wait potentially indefinitely.

Reply:
And while thinking more about this, I realize that maybe the current wording of 5 is perhaps not what is wanted here. In AXI-Stream it would be perfectly fine for a producer to assert `valid` in response to seeing `ready` asserted, provided that the producer also has a way in which it is guaranteed to assert `valid` even if the consumer never asserts `ready`. Is this the rule that is intended here?

Reply:
> Is this the rule that is intended here?

The intent is to exclude (a) comb feedback of `valid` from `ready` (which results in not-well-formed programs) and (b) in general `valid` being dependent on `ready` (which results in deadlocks).

I agree that these two are separate and would have to think about it more.

Reply:
Take for example a producer that asserts `valid` according to these rules (although I'd never write something like this other than for the sake of the example) :

1. Whenever `valid` is low,  a count-down is started on the next cycle.
2. When the count-down reaches the end, the counter stops in this "end-of-countdown" state.
3. If `ready` is high, the counter is set to the "end-of-countdown" state on the next cycle.
4. `valid` is the combinational OR of `ready` and the "end-of-countdown" state.

I think this would be okay in AXI-Stream (now that I've learned that combinational paths between inputs and outputs are not forbidden as in AXI), so the questions is if this producer would be allowed in Amaranth streams.

Reply:
Having re-read the spec:

> A Transmitter is not permitted to wait until TREADY is asserted before asserting TVALID. Once TVALID is asserted, it must remain asserted until the handshake occurs.

and this paragraph:

>  In AXI-Stream it would be perfectly fine for a producer to assert valid in response to seeing ready asserted, provided that the producer also has a way in which it is guaranteed to assert valid even if the consumer never asserts ready. Is this the rule that is intended here?

it is completely unclear to me why what you say is accurate.

Moreover, the following step would still create a combinatorial loop if the receiver tied `ready` to `valid`, so the example you provide does not motivate allowing this functionality.

> 4\. `valid` is the combinational OR of `ready` and the "end-of-countdown" state.

Reply:
My example doesn't wait until TREADY is asserted. It waits until an internal counter expires. But it also asserts TVALID earlier if TREADY becomes asserted. I don't see why the spec would forbid this, even though it's a rather convoluted example.

> Moreover, the following step would still create a combinatorial loop if the receiver tied `ready` to `valid`.

You're right. Then maybe the fact that the AXI-Stream spec doesn't mention that combinatorial paths between input and output are not allowed is a typo rather than an intended choice to make it different from AXI?

Reply:
> My example doesn't wait until TREADY is asserted. It waits until an internal counter expires. But it also asserts TVALID earlier if TREADY becomes asserted.

In my view that counts as "waiting until TREADY" because that's a form of feedback.



> Then maybe the fact that the AXI-Stream spec doesn't mention that combinatorial paths between input and output are not allowed is a typo rather than an intended choice to make it different from AXI?

You only need to prohibit combinatorial paths in one direction (which it does), not in both, to prevent logic loops.

Reply:
In any case I think that your last version https://github.com/amaranth-lang/rfcs/commit/1c9218781c1b8fc6071e3e0dda457cae9c916032 forbids combinatorial loops clearly enough, so the things I'm wondering are more about AXI-Stream than about this RFC.

Reply:
> I don't see why the spec would forbid this, …

The issue is that if you have a source that under some conditions waits for `ready` before setting `valid` and a sink that under some conditions waits for `valid` before setting `ready` and connect them to each other, you've got a potential for a deadlock when they both come into the condition where they're waiting for the other end to make the first move.

To eliminate the possibility of deadlock, one of these must be forbidden, and having `valid` depend on the state of `ready` is the less useful one.

Reply:
@zyp I think @daniestevez's example doesn't have a deadlock. In it, the transmitter always makes forward progress, just at different speed.

Reply:
Ah, yeah. I can't come up with a practical use for code that resembles the example, but since it neither has potential for deadlocks nor combinatorial loops, I don't see a specific reason to disallow it.

> Any form of feedback from `ready` to `valid` is prohibited.

This wording is too strict, as *de*assertion of `valid` very much depends on feedback from `ready`.

Reply:
> Ah, yeah. I can't come up with a practical use for code that resembles the example, but since it neither has potential for deadlocks nor combinatorial loops, I don't see a specific reason to disallow it.

I think having streams be push-only makes understanding the systems they are a part of easier, which to me is significant here.

> 
> > Any form of feedback from `ready` to `valid` is prohibited.
> 
> This wording is too strict, as _de_assertion of `valid` very much depends on feedback from `ready`.

Fixed, thanks. Now it's just the posedge that needs to happen without feedback.